pkg/cloud/external_storage.go

```diff
@@ -182,3 +182,43 @@ type ExternalStorageOptions struct {
 type ExternalStorageConstructor func(
 	context.Context, ExternalStorageContext, cloudpb.ExternalStorage,
 ) (ExternalStorage, error)
+
+// NewExternalStorageAccessor creates an uninitialized ExternalStorageAccessor.
+func NewExternalStorageAccessor() *ExternalStorageAccessor {
+	return &ExternalStorageAccessor{ready: make(chan struct{})}
+}
+
+// ExternalStorageAccessor is a container for accessing the ExternalStorage
+// factory methods once they are initialized. Attempts to access them prior to
+// initialization will block.
+type ExternalStorageAccessor struct {
+	ready   chan struct{}
+	factory ExternalStorageFactory
+	byURI   ExternalStorageFromURIFactory
+}
+
+// Init initializes the ExternalStorageAccessor with the passed factories.
+func (a *ExternalStorageAccessor) Init(
+	factory ExternalStorageFactory, uriFactory ExternalStorageFromURIFactory,
+) error {
+	close(a.ready)
+	a.factory = factory
+	a.byURI = uriFactory
+	return nil
+}
+
+// Open opens an ExternalStorage.
+func (a *ExternalStorageAccessor) Open(
```

Comment:
if some bozo calls Open() before Init(), why block instead of error?

Reply:
/shrug We need to synchronize _anyway_ so blocking is easy. AFAIK, this code is going away anyway (maybe?) as soon as pebble takes over opening/downloading the files (or maybe we should keep it and be adding a hook to pebble to allow passing in the factory implementation that it should be using to open shared file locators, so that cockroach can pass an implementation that then knows how to do nodelocal/etc files? /cc @itsbilal )

Reply:
Yeah, we create the engine before we Init the `ExternalStorageAccessor` the way this code is set up right now. So Pebble will almost certainly call `Open` before Init. I'm not the biggest fan of the synchronization either - is there any reason why we can't create the `ExternalStorageAccessor` closer to where we `CreateEngines` (and ideally before it)? That way we can just pass it into `pebble.Options.Experimental.SharedStorage` and things just fall together.